oci/archive/reader.go

```diff
@@ -0,0 +1,90 @@
+package archive
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/containers/image/v5/internal/tmpdir"
+	"github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/types"
+	"github.com/containers/storage/pkg/archive"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type reader struct {
+	manifest *imgspecv1.Index
+	tempDirOCIRef
+}
+
+// CreateUntarTempDirReader creates the temp directory that keeps the untarred archive from src.
+// The caller should call .Close() on the returned object.
+func CreateUntarTempDirReader(ctx context.Context, src string, sys *types.SystemContext) (*reader, error) {
+	// TODO: This can take quite some time, and should ideally be cancellable using a context.Context.
+	arch, err := os.Open(src)
+	if err != nil {
+		return nil, err
+	}
+	defer arch.Close()
+
+	dst, err := ioutil.TempDir(tmpdir.TemporaryDirectoryForBigFiles(sys), "oci")
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating temp directory")
+	}
+
+	reader := reader{
+		tempDirOCIRef: tempDirOCIRef{tempDirectory: dst},
+	}
+
+	if err := archive.NewDefaultArchiver().Untar(arch, dst, &archive.TarOptions{NoLchown: true}); err != nil {
+		if err := reader.tempDirOCIRef.deleteTempDir(); err != nil {
+			return nil, errors.Wrapf(err, "error deleting temp directory %q", dst)
+		}
+		return nil, errors.Wrapf(err, "error untarring file %q", dst)
+	}
+
+	indexJSON, err := os.Open(filepath.Join(dst, "index.json"))
+	if err != nil {
+		return nil, err
+	}
+	defer indexJSON.Close()
+	reader.manifest = &imgspecv1.Index{}
+	if err := json.NewDecoder(indexJSON).Decode(reader.manifest); err != nil {
+		return nil, err
+	}
+
+	return &reader, nil
+}
+
+// List returns a (name, reference) map for images in the reader
+// the name will be used to determin reference name of the dest image.
+// the ImageReferences are valid only until the Reader is closed.
+func (r *reader) List() (map[string]types.ImageReference, error) {
+	res := make(map[string]types.ImageReference)
+	for _, md := range r.manifest.Manifests {
+		if md.MediaType != imgspecv1.MediaTypeImageManifest && md.MediaType != imgspecv1.MediaTypeImageIndex {
+			continue
+		}
+		refName, ok := md.Annotations[imgspecv1.AnnotationRefName]
+		if !ok {
+			continue
+		}
+		ref, err := layout.NewReference(r.tempDirOCIRef.tempDirectory, refName)
```

Comment:
I don’t think it’s quite that simple — this creates references with the temporary path , which would show up in the UI (and, if anyone ever used it, in `PolicyConfigurationIdentity`).

See how `docker.archiveReference` holds an optional `archiveReader`.

Reply:
Could you elaborate why the `archiveReader` is needed?

Reply:
With this implementation,
```go
ref := /* something like */archive.CreateUntarTempDirReader(… "/home/mitr/my-precious-archive" …).List()["my-tag"]
printf("Now pulling %s", transports.ImageName(ref) // prints Now pulling oci-layout:/var/tmp/ociRANDOM:my-tag
```
and similarly that identity is probably going to show up in error messages. Relating that back to the `my-precious-archive` path is non-obvious.

`PolicyConfigurationIdentity` is theoretically the most severe impact (bypassing/breaking `policy.json` because both the transport and the path is different), in practice probably never hit; the error message aspect is more important.

---

The `archiveReader` approach has another potentially useful (but admittedly not strictly _necessary_) property, that the references created by `reader.List` for accessing individual parts of a compressed archive are, as far as the string syntax shows, exactly the same as the `oci-archive:$file:$tag` references the user can write manually; the `reader` object becomes just a way to optimize accessing many such references with a single tar extraction overhead.